### PR TITLE
add KDPM2DiscreteScheduler

### PR DESCRIPTION
**Pull request by cstueckrath**
and add a force flag for setup_venv.ps1

### CODE REVIEW

File: setup_venv.ps1 [hunk 1]
```diff
@@ -4,6 +4,23 @@ if ($arguments -eq "--update-src"){
...
+	if (Test-Path env:VIRTUAL_ENV) {
+        Write-Host "deactivating..."
+        Deactivate
+    }
+    
+    if (Test-Path .\shark.venv\) {
+        Write-Host "removing and recreating venv..."
+        Remove-Item .\shark.venv -Force -Recurse
+        if (Test-Path .\shark.venv\) {
+            Write-Host "could not remove .\shark-venv - please try running setup_venv.ps1 again!"
```
Comment (cstueckrath) [inline setup_venv.ps1:17]: should be
Write-Host 'could not remove .\shark-venv - please try running "setup_venv.ps1 --force" again!'

File: setup_venv.ps1 [hunk 2]
```diff
@@ -36,6 +53,8 @@ Write-Host $p
 Write-Host "Installing Build Dependencies"
 python -m venv .\shark.venv\
 .\shark.venv\Scripts\activate
+python -m pip install --upgrade pip
+pip install wheel
```
Comment (powderluv) [inline setup_venv.ps1:57]: I would also just remove iree-compiler and runtime here too so it always reinstalls when you run setup_venv.ps1. But can happen later

**Review by powderluv — APPROVED** [review]

File: setup_venv.ps1 (hunk 2 above)
Comment (cstueckrath) [reply setup_venv.ps1:57]: I would like to redo this whole script anyways being object oriented with try-catch phrases and help functions. Also a what-if and verbose is missing.
But that needs some time

**Comment by Shukla-Gaurav** [comment]
@cstueckrath can you please rebase it and also squash the commits to single commit.

**Comment by cstueckrath** [comment]
sure, can you give me instructions for that? I'm not doing this very often...

**Comment by cstueckrath** [comment]
I FUBARed it.
I'll lifeboat my changes, delete my Fork and start with a new PR

**Comment by Shukla-Gaurav** [comment]
> sure, can you give me instructions for that? I'm not doing this very often...
> 
> ```
>  SHARK  git fetch origin
>  SHARK  git rebase origin/main
> Current branch main is up to date.
>  SHARK  git checkout main
> Already on 'main'
> Your branch is up to date with 'origin/main'.
>  SHARK  git merge --squash
> Already up to date. (nothing to squash)
> ```
> 
> Do I have to git rebase -i HEAD~x What should x be?
> 
> Hmm, reading some instructions for all of that now. Can take a while
> 
> Uh, I think I've made it worse...

whenever you make new changes, try to use `git commit --amend`, it will update the same commit instead of creating a new commit. Also force push in this case `git push <repo> <branch> -f`.

**Pull request closed without merging**